docs/Learning-Environment-Design.md

```diff
@@ -83,6 +83,12 @@ training a maze-solving agent, you would probably want to change the maze itself
 for each training episode. Otherwise, the agent would probably on learn to solve
 one, particular maze, not mazes in general.
 
+#### Academy Pitfalls
+
+In some games, it may be desirable to step the environment with the Academy manually via `Academy.Instance.EnvironmentStep()`.
+If this is the case, be sure `Academy.Instance.EnvironmentStep()` is not called directly or indirectly by the agent's
+`CollectObservations()` function as this will cause an infinite loop that prevents that main Academy update function from being called.
```

Comment:
CollectObservations is not the only method that might pose problems. Anything called BY the EnvironmentStep will be problematic. For example: Heuristic and OnActionReceived.

Reply:
Would it be better to guard against this? We could do something like
```
public void EnvironmentStep() {
  if(m_Stepping) {
    throw new UnityException("Infinite recursion prevented!");
  }
  m_Stepping = true;
  // do all the work
  m_Stepping = false;
}
```

Reply:
That works for me. It's up to you both.

Reply:
(with maybe a try/finally around the main body to reset the stepping flag, and/or change the throw to just return)

Reply:
@vincentpierre thoughts?